runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -365,14 +365,166 @@ local function get_range_from_metadata(node, id, metadata)
   return { node:range() }
 end
 
+---@private
+---@param node TSNode
+---@param id integer
+---@param metadata TSMetadata
+---@return Range4[]
+local function get_node_ranges(node, id, metadata, include_children)
+  if include_children then
+    return { get_range_from_metadata(node, id, metadata) }
+  end
+
+  local ranges = {} ---@type Range4[]
+
+  local srow, scol, erow, ecol = node:range()
+
+  for i = 0, node:named_child_count() - 1 do
+    local child = node:named_child(i)
+    local child_srow, child_scol, child_erow, child_ecol = child:range()
+    if child_srow > srow or child_scol > scol then
+      table.insert(ranges, { srow, scol, child_srow, child_scol })
+    end
+    srow = child_erow
+    scol = child_ecol
+  end
+
+  if erow > srow or ecol > scol then
+    table.insert(ranges, { srow, scol, erow, ecol })
+  end
+
+  return ranges
+end
+
+---@private
+---@param t table<integer,table<string,table<integer,table>>>
+---@param tree_index integer
+---@param pattern integer
+---@param lang string
+---@param combined boolean
+---@param ranges Range4[]
+local function add_injection(t, tree_index, pattern, lang, combined, ranges)
+  assert(type(lang) == 'string')
+
+  -- Each tree index should be isolated from the other nodes.
+  if not t[tree_index] then
+    t[tree_index] = {}
+  end
+
+  if not t[tree_index][lang] then
+    t[tree_index][lang] = {}
+  end
+
+  -- Key this by pattern. If combined is set to true all captures of this pattern
+  -- will be parsed by treesitter as the same "source".
+  -- If combined is false, each "region" will be parsed as a single source.
+  if not t[tree_index][lang][pattern] then
+    t[tree_index][lang][pattern] = { combined = combined, regions = {} }
+  end
+
+  table.insert(t[tree_index][lang][pattern].regions, ranges)
+end
+
+---@private
+---@param node TSNode
+---@param source integer|string
+---@param metadata table
+---@return string
+local function get_node_text(node, source, metadata)
```

Comment:
same -- we already have a `vim.treesitter.get_node_text()`

Reply:
This is just to shut up the LSP and avoid hacky annotations.

Reply:
worth a TODO comment then, otherwise you'll get hop-on comments (like this one)